zebra-consensus/src/block/subsidies.rs

```diff
@@ -0,0 +1,256 @@
+
+use zebra_chain::{
+    block::Height,
+    parameters::{Network, NetworkUpgrade::*},
+};
+
+use crate::parameters::{fs, Params};
+
+#[cfg(test)]
+use color_eyre::Report;
+
+fn slow_start_shift() -> Height {
+    Height(Params::SLOW_START_INTERVAL / 2)
+}
+
+fn slow_start_rate() -> u32 {
+    Params::MAX_BLOCK_SUBSIDY / Params::SLOW_START_INTERVAL
+}
+
+fn halving(height: Height, network: Network) -> u32 {
+    if height < slow_start_shift() {
+        return 0;
+    }
+    Blossom
+        .activation_height(network)
+        .filter(|&blossom_height| height >= blossom_height)
+        .map(|blossom_height| {
+            (blossom_height.0 - slow_start_shift().0) * Params::BLOSSOM_POW_TARGET_SPACING_RATIO
+                + height.0
+                - blossom_height.0
+        })
+        .map(|scaled_halvings| scaled_halvings / Params::POST_BLOSSOM_HALVING_INTERVAL)
+        .unwrap_or_else(|| (height.0 - slow_start_shift().0) / Params::PRE_BLOSSOM_HALVING_INTERVAL)
+}
+
+pub fn block_subsidy(height: Height, network: Network) -> u32 {
+    if height < slow_start_shift() {
+        slow_start_rate() * height.0
+    } else if slow_start_shift() <= height && height < Height(Params::SLOW_START_INTERVAL) {
+        slow_start_rate() * (height.0 + 1)
+    } else {
+        Blossom
+            .activation_height(network)
+            .filter(|&blossom_height| {
+                height < blossom_height && Params::SLOW_START_INTERVAL <= height.0
+            })
+            .map(|_| Params::MAX_BLOCK_SUBSIDY >> halving(height, network))
+            .unwrap_or_else(|| {
+                ((Params::MAX_BLOCK_SUBSIDY / Params::BLOSSOM_POW_TARGET_SPACING_RATIO
+                    * 2u32.pow(halving(height, network))) as f32)
+                    .floor() as u32
+            })
+    }
+}
+
+fn founders_reward(height: Height, network: Network) -> u32 {
+    if halving(height, network) < 1 {
+        (block_subsidy(height, network) as f32 * Params::FOUNDERS_FRACTION).floor() as u32
+    } else {
+        0
+    }
+}
+
+fn funding_stream(height: Height, network: Network, receiver: fs::Receiver) -> u32 {
+    if height.0 >= Params::CANOPY_ACTIVATION_HEIGHT
+        && fs::mainnet::START_HEIGHT <= height.0
+        && height.0 < fs::mainnet::END_HEIGHT
+    {
+        (block_subsidy(height, network) as f32
+            * (fs::mainnet::numerator(receiver) as f32 / fs::mainnet::DENOMINATOR as f32))
+            .floor() as u32
+    } else {
+        0
+    }
+}
+
+pub fn miner_subsidy(height: Height, network: Network) -> u32 {
+    let mut funding_streams: u32 = 0;
+    funding_streams += funding_stream(height, network, fs::Receiver::ECC);
+    funding_streams += funding_stream(height, network, fs::Receiver::ZF);
+    funding_streams += funding_stream(height, network, fs::Receiver::MG);
+
+    block_subsidy(height, network) - founders_reward(height, network) - funding_streams
+}
+
+#[test]
+fn test_halving() -> Result<(), Report> {
```

Comment:
These tests should be in a `tests` submodule annotated with `#[cfg(test)]` so they're only compiled when we're running tests.

Reply:
Moved at https://github.com/ZcashFoundation/zebra/pull/1051/commits/15409e60a10c24ca6b013fc253b9744d44e7624b
Thanks.